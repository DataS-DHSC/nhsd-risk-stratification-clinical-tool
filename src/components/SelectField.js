import React from 'react';
import { Select } from 'nhsuk-react-components';
import PropTypes from 'prop-types';

const SelectField = ({ field, groupName, prevData, required, setField }) => {
  const { label, options, hint } = field;

  return (
    <Select
      className="nhsuk-select"
      id={groupName}
      name={groupName}
      disabled={field.disabled}
      hint={hint}
      label={label}
      defaultValue={
        prevData[groupName] ||
        (options[0].options[0].name === 'please-select'
          ? ''
          : options[0].options[0].name.toString())
      }
      required={required}
      onChange={(e) => setField(groupName, e.target.value)}
    >
      {options.map((optgroup) => {
        const subOptions = optgroup.options.map((option) => (
          <option
            key={`option-${option.name}`}
            value={
              option.name === 'please-select' ? '' : option.name.toString()
            }
          >
            {option.label}
          </option>
        ));

        return optgroup.label !== '' ? (
          <optgroup label={optgroup.label} key={`optgroup-${optgroup.label}`}>
            {subOptions}
          </optgroup>
        ) : (
          subOptions
        );
      })}
    </Select>
  );
};

SelectField.propTypes = {
  field: PropTypes.shape().isRequired,
  prevData: PropTypes.shape(),
  groupName: PropTypes.string.isRequired,
  required: PropTypes.bool,
  setField: PropTypes.func.isRequired,
};

SelectField.defaultProps = {
  prevData: {},
  required: false,
};

export default SelectField;
