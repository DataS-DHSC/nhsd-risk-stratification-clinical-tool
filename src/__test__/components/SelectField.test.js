import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SelectField from '../../components/SelectField';

const setupDefaultProps = (defaultFieldValue) => {
  const props = {
    field: {
      label: 'testLabel',
      type: 'select',
      default: defaultFieldValue,
      options: [
        {
          label: '',
          options: [{ name: 'please-select', label: 'Please select' }],
        },
        {
          label: 'subGroup1',
          options: [
            { name: 'sub1value1', label: 'Value11' },
            { name: 'sub1value2', label: 'Value12' },
          ],
        },
        {
          label: 'subGroup2',
          options: [
            { name: 'sub2value1', label: 'Value21' },
            { name: 'sub2value2', label: 'Value22' },
          ],
        },
      ],
    },
    groupName: 'testValueId',
    prevData: {},
    setField: jest.fn(),
  };

  return props;
};

describe('SelectField', () => {
  it('should match snapshot when there is no default nor prevData value', () => {
    const props = { ...setupDefaultProps(undefined), prevData: {} };
    const container = render(<SelectField {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should take its value from prevData rather than the default', () => {
    const props = {
      ...setupDefaultProps('sub1value2'),
      prevData: { testValueId: 'sub1value1' },
    };

    const container = render(<SelectField {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should take the default value when the field is not in prevData', () => {
    const props = { ...setupDefaultProps('sub1value2'), prevData: {} };
    const container = render(<SelectField {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should call setField on value change', () => {
    const setField = jest.fn();

    const props = {
      ...setupDefaultProps('sub1value1'),
      setField,
    };

    const { getByRole } = render(<SelectField {...props} />);

    userEvent.selectOptions(getByRole('combobox'), ['sub1value2']);

    expect(setField).toHaveBeenCalledWith('testValueId', 'sub1value2');
  });
});
