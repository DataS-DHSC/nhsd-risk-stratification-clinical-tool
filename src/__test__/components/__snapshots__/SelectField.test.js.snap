// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField should match snapshot when there is no default nor prevData value 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="testValueId"
      id="testValueId--label"
    >
      testLabel
    </label>
    <select
      aria-labelledby="testValueId--label"
      class="nhsuk-select nhsuk-select"
      id="testValueId"
      name="testValueId"
    >
      <option
        selected=""
        value=""
      >
        Please select
      </option>
      <optgroup
        label="subGroup1"
      >
        <option
          value="sub1value1"
        >
          Value11
        </option>
        <option
          value="sub1value2"
        >
          Value12
        </option>
      </optgroup>
      <optgroup
        label="subGroup2"
      >
        <option
          value="sub2value1"
        >
          Value21
        </option>
        <option
          value="sub2value2"
        >
          Value22
        </option>
      </optgroup>
    </select>
  </div>
</DocumentFragment>
`;

exports[`SelectField should take its value from prevData rather than the default 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="testValueId"
      id="testValueId--label"
    >
      testLabel
    </label>
    <select
      aria-labelledby="testValueId--label"
      class="nhsuk-select nhsuk-select"
      id="testValueId"
      name="testValueId"
    >
      <option
        value=""
      >
        Please select
      </option>
      <optgroup
        label="subGroup1"
      >
        <option
          selected=""
          value="sub1value1"
        >
          Value11
        </option>
        <option
          value="sub1value2"
        >
          Value12
        </option>
      </optgroup>
      <optgroup
        label="subGroup2"
      >
        <option
          value="sub2value1"
        >
          Value21
        </option>
        <option
          value="sub2value2"
        >
          Value22
        </option>
      </optgroup>
    </select>
  </div>
</DocumentFragment>
`;

exports[`SelectField should take the default value when the field is not in prevData 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="testValueId"
      id="testValueId--label"
    >
      testLabel
    </label>
    <select
      aria-labelledby="testValueId--label"
      class="nhsuk-select nhsuk-select"
      id="testValueId"
      name="testValueId"
    >
      <option
        selected=""
        value=""
      >
        Please select
      </option>
      <optgroup
        label="subGroup1"
      >
        <option
          value="sub1value1"
        >
          Value11
        </option>
        <option
          value="sub1value2"
        >
          Value12
        </option>
      </optgroup>
      <optgroup
        label="subGroup2"
      >
        <option
          value="sub2value1"
        >
          Value21
        </option>
        <option
          value="sub2value2"
        >
          Value22
        </option>
      </optgroup>
    </select>
  </div>
</DocumentFragment>
`;
